File: task_2/task_2.java
import java.util.*;

public class Main {

    public static boolean limitations(String str) { // ПРОВЕРКА

        String symbols = "qwertyuiopasdfghjklzxcvbnm0123456789.+";
        String zapret = "&=+<>,_`-";
        String strName = "";
        String strDomen = "@";

        if (!str.contains("@")) { // если нет @
            return false;
        }

        int pos = str.indexOf('@');
        strDomen += str.substring(pos + 1); // домен

        if (!strDomen.contains(".") || strDomen.length() < 4) {
            return false;
        }

        // проверка запрещённых символов в домене
        for (char i : strDomen.toCharArray()) {
            boolean prov = true;
            for (char j : zapret.toCharArray()) {
                if (i == j) {
                    prov = false;
                    break;
                }
            }
            if (!prov) {
                return false;
            }
        }

        // имя пользователя
        for (char c : str.toCharArray()) {
            if (c == '@') break;
            strName += c;
        }

        if (strName.length() < 6 || strName.length() > 30) {
            return false;
        }

        // проверка на ".."
        for (int i = 0; i < str.length() - 1; i++) {
            if (str.charAt(i) == '.' && str.charAt(i + 1) == '.') {
                return false;
            }
        }

        // проверка допустимых символов
        for (char c : strName.toCharArray()) {
            boolean prov = false;
            for (char j : symbols.toCharArray()) {
                if (c == j) {
                    prov = true;
                    break;
                }
            }
            if (!prov) {
                return false;
            }
        }

        if (strName.charAt(0) == '.' || strName.charAt(strName.length() - 1) == '.') {
            return false;
        }

        return true;
    }

    public static String normalEmail(String str) { // нормализация
        int pos = str.indexOf('@');
        String name = str.substring(0, pos);
        String domen = str.substring(pos);

        StringBuilder cleanName = new StringBuilder();

        int plusPos = name.indexOf('+');
        if (plusPos != -1) {
            name = name.substring(0, plusPos);
        }

        for (char c : name.toCharArray()) {
            if (c != '.') {
                cleanName.append(c);
            }
        }

        return cleanName.toString() + domen;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Set<String> emails = new HashSet<>();

        String email = "";

        while (!email.equals("end")) {
            email = scanner.next();

            if (!email.equals("end") && !limitations(email)) {
                System.out.println("EROR!!!");
            } else {
                if (!email.equals("end")) {
                    email = normalEmail(email);
                    emails.add(email);
                }
            }
        }

        System.out.println("Уникальных адресов " + emails.size());
    }
}

File: task_2/task_2.cpp
#include<iostream>
#include<string>
#include<set>
using namespace std;

bool limitations (const string& str){ //ПРОВЕРКА

    string symbols = "qwertyuiopasdfghjklzxcvbnm0123456789.+"; // допустимые символы 
    string zapret = "&=+<>,_`-"; // запрещённые символы
    string strName = ""; // имя пользователя
    string strDomen = "@"; // @ + доменное имя

    if (str.find('@') == string::npos){ //если нет @
        return false;
    }

    size_t pos = str.find('@'); 
    strDomen += str.substr(pos + 1); //находим доменное имя

    if (strDomen.find('.') == string::npos || strDomen.size()<4){ //проверяем нормальное ли доменное имя
        return false;
    }

    for (char i : strDomen){ // проверка на запрещённые символы в домене
        bool prov = true;
        for (char j : zapret){
            if (i == j){
                prov = false;
                break;
            }
        }
        if (prov == false){
            return false;
        }
    }
    
    for (char sN : str){ // находим имя пользователя
        if (sN == '@'){
            break;
        }
        else{
            strName+=sN;
        }
    }
    if (strName.size() <6 || strName.size()>30){ // проверка длины имени 
        return false;
    }

    for (int q = 0; q <str.size()-1; q++){ // проверка на несколько точек подряд
        if (str[q] == '.' && str[q+1] == '.'){
            return false;
        }
    }

    for (char i : strName){ // проверка на допустимые символы в имени 
        bool prov = false;
        for (char j : symbols){
            if (i == j){
                prov = true;
                break;
            }
        }
        if (prov == false){
            return false;
        }
    }

    if (strName[0]=='.' || strName[strName.size() -1]== '.'){  //проверка на то, что имя не начинается и не заканчивается на точку
        return false;
    }
    return true;
}

string normalEmail(const string& str){ // нормализируем, убираем всё после +, удаляем не нужные точки 
    size_t pos = str.find('@');
    string name = str.substr(0, pos);
    string domen = str.substr(pos);
    string cleanName = ""; // строка без точек

    size_t plusPos = name.find('+');
    if (plusPos != string::npos){ // считаем имя только до +
        name =name.substr(0, plusPos);
    }

    for (char c : name){ // удаляем все лишние точки 
        if (c != '.'){
            cleanName += c;
        }
    }

    return cleanName+domen;

}



int main(){
    string email;
    set<string> emails;
    bool limit = false;
    while (email!="end"){ // вводим email пока не введём end
        cin >> email; 
        if (limitations(email) == false && email !="end"){
            cout <<"ERROR!!!" <<endl;
        } else {
            if (email !="end"){
                email = normalEmail(email);
                emails.insert(email);
            }
        }
        
    }

    cout << "Уникальных адресов " << emails.size() << endl;

}